### ISSUE TITLE
I want to stop a continuous async recognition immediately

### ISSUE THREAD

**Issue by davehorton** [Azure-Samples/cognitive-services-speech-sdk#1388]
I have a scenario where I am using `StartContinuousRecognitionAsync` to start a recognition.  Some time later I want to stop the recognition, but in this scenario I am not interested in final results and do not want to block my thread for the ~10 secs it is typically taking for `StopContinuousRecognitionAsync().get();` to complete (for some reason almost always I have to wait for a timeout on the server side).

How can I stop the recognizer instantly?

**Comment by jhakulin** [comment]
@davehorton Please check if this https://github.com/Azure-Samples/cognitive-services-speech-sdk/issues/1363#issuecomment-1006011016 would provide improvement. We are working on improving the stop behavior of Speech SDK.

**Comment by davehorton** [comment]
I will try the "fire and forget" approach.  I believe I did try that earlier and somehow I still encountered a block when I destroyed the recognizer....will retry and report

**Comment by davehorton** [comment]
the "fire and forget" approach still blocks for me.  As far as I can tell, when I call `m_recognizer->StopContinuousRecognitionAsync()` it is returning a future which, regardless if I ignore it, when it goes out of scope and is destroyed causes the thread to block for 10 secs.

This is forcing me into writing much more complex (and brittle) code as compared to the integrations I've previously done with AWS and google streaming transcribers.  As it stands now, it looks like I will have to create some sort of "reaper" facility which will have to manage these synthesizers that I want to discard by spinning up a new thread for each to wait out its server side timeout.  :(

**Comment by trrwilson** [comment]
@davehorton the gotcha is that there are a variety of possible "stop" behaviors (abort immediately, end audio immediately but wait for a result, wait for the next natural break and finish are common ones) and `StopContinuousRecognitionAsync` is closest to that last one: it dutifully attempts to finish the audio that's in the middle of being used and then waits for a result.

1.20 will include an improvement specific to how `PushAudioInputStream` behaves with Stop: currently, if you call `StopContinuousRecognitionAsync` without the input stream subsequently providing up to 100ms or so of new audio once that stop request begins, the SDK treats that as a wait on the data source -- which is where that unwieldy 10s comes from. With the new release, `Stop` will immediately flush as if an end of stream had been reached.

That'll get towards that middle one, which is generally a wait on the order of dozens to a couple hundred milliseconds to wait for a final result. This is some timed output from a quick test using .Close() to stop in the middle of a "hello world, this is a test" utterance provided via a stream:

```
[56:49.898] RecognizingSpeech:     a19bb3ce1dc948729a3e620c1d3a6fb5: hello
[56:50.052] Calling PushAudioInputStream::Close (Audio caught up)
[56:50.276] RecognizingSpeech:     a19bb3ce1dc948729a3e620c1d3a6fb5: hello world
[56:50.292] RecognizedSpeech:     a19bb3ce1dc948729a3e620c1d3a6fb5: Hello world.
[56:50.295]         Canceled:     a19bb3ce1dc948729a3e620c1d3a6fb5: (EndOfStream):
```

That's a little under 250ms between calling `.Close()` and receiving the `Canceled` event with `EndOfStream` that indicates the recognizer is idle and ready for reuse--not instant, but a lot better than 10000+ms!

We unfortunately don't yet have a standalone "abort" analogue apart from doing the above and then throwing away/ignoring the instance.  I'd love for us to have one and it's on our backlog.

**Comment by davehorton** [comment]
Thanks, this sounds good and really appreciate the update and insight!

**Comment by adammpolak** [comment]
> 

@trrwilson does the Stop response provide the full transcript of everything that has been said (since continuous async recognition began)?

**Comment by dargilco** [comment]
From the [1.20 release notes](https://docs.microsoft.com/azure/cognitive-services/speech-service/releasenotes?tabs=speech-sdk):
C++, C#, Java, JavaScript, Objective-C, and Swift: Fix to remove a 10-second delay while stopping a speech recognizer that uses a PushAudioInputStream. This is for the case where no new audio is pushed in after StopContinuousRecognition is called (GitHub issues 1318, 331)

**Comment by dargilco** [comment]
@adammpolak We always send a final result ("Recognized" event) based on whatever audio received by the time StopContinuousRecognitionAsync was called. So if you're in the middle of saying "this is a test" you may get "This is." as a final recognition result ("recognized" event) if you called Stop in the middle of the sentence. 

The Recognized event will give you recognition results since the previous Recognized event (not since your started continuous recognition). In continuous recognition, you will get multiple Recognized events, after each pause in the audio (after each sentence).

**Comment by dargilco** [comment]
Closing as it seems the question was answered. Feel free to tag me for follow up questions here, or open a new issue. Thanks!

**Comment by AlexandrosParamythisContexity** [comment]
Sorry for asking on a closed issue, but I thought this is most likely the best place to get the attention of the right people:  what is the situation with the Python client library? In our tests it seems to be affected by the 10-second freeze, no matter what we try. The release notes pointed at by @dargilco do not include python, and the issue is also not mentioned in the 1.21 release notes. Thanks in advance for any feedback!

**Comment by aprotopapas** [comment]
> Sorry for asking on a closed issue, but I thought this is most likely the best place to get the attention of the right people: what is the situation with the Python client library? In our tests it seems to be affected by the 10-second freeze, no matter what we try. The release notes pointed at by @dargilco do not include python, and the issue is also not mentioned in the 1.21 release notes. Thanks in advance for any feedback!

Did you ever get a response to this?

**Comment by dandiep** [comment]
@dargilco I attached a test showing the 10s pause when using threads in #2629. Any chance you could take a look?
